### PR TITLE
WINC-715: Introduce Windows services ConfigMap schema

### PR DESCRIPTION
**Pull request by saifshaikh48**
This PR covers the data representation of a service ConfigMap which provides WICD with the specifications for each Windows service that must be created on a Windows instance.

- Services ConfigMap follows the naming pattern `windows-services-<WMCOMajorVersion>-<WMCOMinorVersion>-<WMCOPatchVersion>-<Commit>`

### CODE REVIEW

File: pkg/configmap/services/services.go [hunk 1]
```diff
@@ -85,6 +87,29 @@ func NewServicesConfigMapData(services *[]Service, files *[]FileInfo) (*services
 	return cmData, nil
 }
 
+// GenerateWindowsServicesConfigMap creates an immutable service ConfigMap which provides WICD with the specifications
+// for each Windows service that must be created on a Windows instance.
+func GenerateWindowsServicesConfigMap(name, namespace string) *core.ConfigMap {
```
Comment (sebsoto) [inline pkg/configmap/services/services.go:92]: This commit seems like it belongs in the other PR, where this would actually be used

File: test/e2e/create_test.go [hunk 2]
```diff
@@ -15,19 +15,19 @@ import (
 	"github.com/pkg/errors"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
-	"k8s.io/api/core/v1"
+	core "k8s.io/api/core/v1"
```
Comment (sebsoto) [inline test/e2e/create_test.go:18]: this is an unrelated change

File: pkg/configmap/services/services.go [hunk 3]
```diff
@@ -57,6 +68,119 @@ type FileInfo struct {
...
+
+	services := &[]Service{}
+	files := &[]FileInfo{}
+
+	value, ok := data[servicesKey]
+	if !ok {
+		return nil, errors.Errorf("expected `%s` key to exist", servicesKey)
+	}
+	if err := json.Unmarshal([]byte(value), services); err != nil {
+		return nil, errors.Wrapf(err, "`%s` must represent list of Service", servicesKey)
```
Comment (sebsoto) [inline pkg/configmap/services/services.go:128]: a list of Services

Comment (sebsoto) [reply pkg/configmap/services/services.go:128]: could just return error without wrapping it as i dont think you're adding much info here

File: pkg/configmap/services/services.go [hunk 4]
```diff
@@ -57,6 +68,119 @@ type FileInfo struct {
...
+	}
+
+	value, ok = data[filesKey]
+	if !ok {
+		return nil, errors.Errorf("expected `%s` key to exist", filesKey)
+	}
+	if err := json.Unmarshal([]byte(value), files); err != nil {
+		return nil, errors.Wrapf(err, "`%s` must represent list of FileInfo", filesKey)
```
Comment (sebsoto) [inline pkg/configmap/services/services.go:136]: same as above

File: pkg/configmap/services/services.go [hunk 5]
```diff
@@ -57,6 +68,119 @@ type FileInfo struct {
...
+
+	for _, bootstrapSvc := range boostrapServices {
+		for _, dependency := range bootstrapSvc.Dependencies {
+			for _, nonBootstrapSvc := range nonBoostrapServices {
+				if dependency == nonBootstrapSvc.Name {
+					return errors.Errorf("bootstrap service %s cannot depend on non-boostrap service %s",
+						bootstrapSvc.Name, nonBootstrapSvc.Name)
+				}
+			}
+		}
```
Comment (sebsoto) [inline pkg/configmap/services/services.go:175]: I think this should be moved into a separate function that takes in a bootstrap service and the slice of nonBootstrap services. The logic is good, just a bit hard to read. example
```
func hasDependency(s Service, possibleDependencies []Service) bool{
		for _, dependency := range s.Dependencies {
			for _, possibleDependency := range possibleDependencies {
				if dependency == possibleDependency.Name {
					return true
				}
			}
			return false
}
```

File: pkg/configmap/services/services.go [hunk 6]
```diff
@@ -57,6 +68,119 @@ type FileInfo struct {
...
+	}
+	return nil
+}
+
+// validatePriorities ensures that each service that has the bootstrap flag set as true has a higher priority than all
+// non-bootstrap services. There should be no overlap in the priorities of bootstrap services and controller services.
+func validatePriorities(services *[]Service) error {
+	// sort services to ascending priority, bootstrap services towards the front of slice
```
Comment (sebsoto) [inline pkg/configmap/services/services.go:183]: services to ascending -> services in ascending

File: pkg/configmap/services/services.go [hunk 7]
```diff
@@ -57,6 +68,119 @@ type FileInfo struct {
...
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// validatePriorities ensures that each service that has the bootstrap flag set as true has a higher priority than all
+// non-bootstrap services. There should be no overlap in the priorities of bootstrap services and controller services.
+func validatePriorities(services *[]Service) error {
```
Comment (sebsoto) [inline pkg/configmap/services/services.go:182]: I could be wrong, but aren't slices already a pointer?
passing in a pointer to slice is redundant

**Review by sebsoto — CHANGES_REQUESTED** [review]

File: pkg/configmap/services/services.go (hunk 1 above)
Comment (saifshaikh48) [reply pkg/configmap/services/services.go]: True. I kept going back and forth whether include it or not haha

File: pkg/configmap/services/services.go (hunk 5 above)
Comment (saifshaikh48) [reply pkg/configmap/services/services.go]: Thanks, I used your snippet as is.

File: pkg/configmap/services/services.go [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+// ServicesConfigMapName is the name of the ConfigMap detailing service configuration for a specific WMCO version
+var ServicesConfigMapName string
+
+// init runs once, initializing global variables
+func init() {
+	ServicesConfigMapName = servicesConfigMapName()
+}
+
+// servicesConfigMapName returns the ConfigMap with the naming scheme:
+// windows-services-<MajorVersion>-<MinorVersion>-<PatchVersion>-<CommitHash>
```
Comment (jrvaldes) [inline pkg/configmap/services/services.go:188]:
```suggestion
// windows-services-<WMCO Full Version>
```


The WMCO full version schema is maintained [elsewhere](https://github.com/openshift/windows-machine-config-operator/blob/master/build/build.sh#L11). Here, we just consume it.

File: pkg/configmap/services/services.go [hunk 9]
```diff
@@ -14,6 +14,49 @@ func init() {
 	ServicesConfigMapName = servicesConfigMapName()
```
Comment (jrvaldes) [inline pkg/configmap/services/services.go:18]: Consider squashing this commit with the previous one. (https://github.com/openshift/windows-machine-config-operator/pull/1004/commits/f523a455c16269b5304ea47d1b1b3c534c7888f9)

File: build/build.sh [hunk 10]
```diff
@@ -8,7 +8,7 @@ set -o pipefail
 add_git_data_to_version() {
   local WMCO_SEMVER=$1
   GIT_COMMIT=$(git rev-parse --short HEAD)
-  FULL_VERSION="${WMCO_SEMVER}+${GIT_COMMIT}"
+  FULL_VERSION="${WMCO_SEMVER}-${GIT_COMMIT}"
```
[the remark this answers is not in the document]
Comment (saifshaikh48) [reply build/build.sh:11]: Given the change is directly relevant to the work in this PR, I would like to keep this change here. As far as testing, the rest of the changes will have no effect on existing functionality (the CM data schema is not plugged in to reconcile logic yet) so we'll know exactly what's causing issues if they arise.

File: pkg/servicescm/servicescm.go [hunk 11]
```diff
@@ -57,6 +68,123 @@ type FileInfo struct {
...
+	for _, bootstrapSvc := range boostrapServices {
+		if hasDependency(bootstrapSvc, nonBoostrapServices) {
+			return errors.Errorf("bootstrap service %s cannot depend on non-boostrap service", bootstrapSvc.Name)
+		}
+	}
+	return nil
+}
+
+// hasDependency checks if a service is dependent on any services in the given slice
+func hasDependency(s Service, possibleDependencies []Service) bool {
```
Comment (aravindhp) [inline pkg/servicescm/servicescm.go:147]: This can be method of the `Service` struct.

**Review by aravindhp — CHANGES_REQUESTED** [review]
Thanks for working on this, @saifshaikh48. Please address my comments. 

I found this a bit hard to review as it is unclear how the functions being introduced here are going to be used which makes me wonder if this is the correct way to split up the PRs.

File: pkg/servicescm/servicescm.go [hunk 12]
```diff
@@ -0,0 +1,21 @@
+package servicescm
```
Comment (alinaryan) [inline pkg/servicescm/servicescm.go:1]: For simplicity, how about referring to this just as `services` in the package name, commit message etc?

**Review by alinaryan — CHANGES_REQUESTED** [review]
Nice work so far, just one comment:

File: pkg/servicescm/servicescm.go [hunk 13]
```diff
@@ -57,6 +68,123 @@ type FileInfo struct {
...
+	cmData := &servicesCMData{*services, *files}
+	if err := cmData.validate(); err != nil {
+		return nil, errors.Wrap(err, "unable to create services ConfigMap data object")
+	}
+	return cmData, nil
+}
+
+// Parse converts ConfigMap data into the objects representing a Windows services ConfigMap schema
+// Returns error if the given data is invalid in structure or contents.
+func Parse(data map[string]string) (*servicesCMData, error) {
```
[the remark this answers is not in the document]
Comment (selansen) [reply pkg/servicescm/servicescm.go:90]: why do we need hash Map here? won't a simple string slice work?
